### PR TITLE
Add migration guide [ECR-2763]

### PR DESCRIPTION
**Pull request by dmitry-timofeev**
## Overview

---
See: https://jira.bf.local/browse/ECR-2763

### Definition of Done

- [x] **Javadocs links are added**
- [x] **Doc links are added** 
- [x] There are no TODOs left in the code — only non-blocking one remains
- [x] Changelog is updated if needed (in case of notable or breaking changes)
- [x] The continuous integration build passes

### CODE REVIEW

File: doc/Migration_guide_0.4.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+Update the clients to submit transactions in the new format, introduced in Exonum 0.10.
+The most notable change in the format is that the public key of the transaction author 
+must always be included in the transaction message. This allows the framework to verify 
+the message signature with no conversion to an executable transaction.
+
+If the client application is in Java, use `com.exonum.binding.common.message.TransactionMessage`;
+if it is in Java Script, use the latest version of the JS light client.
+
+The light client shall submit transactions directly to the core instead of a custom endpoint
+in the service controller. <!-- TODO: which core endpoint? -->
```
Comment (vitvakatu) [inline doc/Migration_guide_0.4.md]: `<public api IP address and port>/api/explorer/v1/transactions`

File: doc/Migration_guide_0.4.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+from the passed context: `TransactionContext#getFooKey`.
+
+### Update `Transaction`s
+The following methods are **removed** from `Transaction` interface:
+  - `isValid`. In v0.4 the framework verifies the cryptographic signature itself, using the
+  public key from the message. If your service performs any additional checks, 
+  move them to the constructor of your transaction or to `TransactionConverter`, whichever 
+  works best.
+  - `info`. It is no longer possible to serialize the transaction in JSON in core endpoints.
+  <!-- ^ Or shall we keep it? -->
```
Comment (vitvakatu) [inline doc/Migration_guide_0.4.md]: Now I think that we actually need this method. Core uses `serde::Serialize` implementation to represent `Transaction` in JSON, but we cannot directly implement `Serialize` on our `TransactionProxy` - so we need some helper method that will return optional JSON description

File: doc/Migration_guide_0.4.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  - `getMessage`. Transactions are no longer required to store the corresponding 
+  `TransactionMessage`, which makes it easier to instantiate them in unit or integration tests.
+  If your transaction code needs a whole message of this or _any_ other transaction,
+  use `com.exonum.binding.blockchain.Blockchain.getTxMessages`.
+  - `hash`. You can access the SHA-256 hash of this transaction message in `Transaction#execute`
+  from the passed context: `TransactionContext#getFooHash`. As a reminder, it uniquely identifies
```
Comment (vitvakatu) [inline doc/Migration_guide_0.4.md]: getFooHash?

File: doc/Migration_guide_0.4.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+them, e.g., in their event handlers. But use with care — each node has its own service keypair.
+<!-- TODO: Some links would be welcome --> 
+
+## Use example
+See how the cryptocurrency service has changed since the last release to understand what
+changes might be needed in your service: <!-- TODO: diff link -->
+
+## Questions?
+
+In case of any questions, feel free to ask a question in Gitter or open an issue on Github.
```
Comment (vitvakatu) [inline doc/Migration_guide_0.4.md]: Link to gitter?

File: doc/Migration_guide_0.4.md (hunk 2 above)
Comment (MakarovS) [reply doc/Migration_guide_0.4.md]: I think we need to be able to serialize our transactions in JSON.

File: doc/Migration_guide_0.4.md (hunk 3 above)
Comment (dmitry-timofeev) [reply doc/Migration_guide_0.4.md]: It has not landed yet :upside_down_face:

File: doc/Migration_guide_0.4.md (hunk 2 above)
Comment (dmitry-timofeev) [reply doc/Migration_guide_0.4.md]: Discussed today that we will keep `info` as is, but also consider pulling it into the framework code (e.g., by using a reflective serializer such as Gson to convert a tx instance to JSON). As it won't _always_ work as one might expect, we must also design a way to extend/override this mechanism.

File: doc/Migration_guide_0.4.md [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+### Use the new message format
+
+EJB 0.4 accepts the transactions via a system endpoint and passes the payload to the `Service`
+implementation to convert into an executable transaction. The `TransactionConverter` now accepts
+a `RawTransaction` that includes only the transaction identifiers (service id and transaction id)
+and its payload — the serialized transaction parameters (equivalent to `Message#getBody` from v0.3).
+
+Also, consider removing the public key of the transaction author from its _payload_ because
+it is now included in a standard field of the message, and is accessible in `Transaction#execute`
+from the passed context: `TransactionContext#getFooKey`.
```
Comment (bullet-tooth) [inline doc/Migration_guide_0.4.md]:
```suggestion
from the passed context: `TransactionContext#getAuthorPk`.
```

File: doc/Migration_guide_0.4.md (hunk 3 above)
Comment (bullet-tooth) [reply doc/Migration_guide_0.4.md]: `getTransactionMessageHash`

**Comment by dmitry-timofeev** [comment]
> creating a path for cryptocurrency demo as a migration example.

What do you mean by that? A patch with changes to the demo?

**Comment by dmitry-timofeev** [comment]
> A patch with changes to the demo?

I wonder if Github can show a diff between two revisions for a certain _path_ (i.e., not a diff for all files).

File: doc/Migration_guide_0.4.md [hunk 6]
```diff
@@ -77,7 +77,12 @@ them, e.g., in their event handlers. But use with care — each node has its own
 
 ## Use example
 See how the cryptocurrency service has changed since the last release to understand what
-changes might be needed in your service: <!-- TODO: diff link -->
+changes might be needed in your service: <!-- TODO: diff link doesn't provide file filters -->
+
+```sh
+$ git clone git@github.com:exonum/exonum-java-binding.git
+$ git diff v0.3 v0.4 exonum-java-binding-cryptocurrency-demo/
```
Comment (bullet-tooth) [inline doc/Migration_guide_0.4.md]: 👍

File: doc/Migration_guide_0.4.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+to access a `Fork` to modify the database state, as previously; and some information about 
+the corresponding transaction message: its hash and the public key of the author.
+
+### Update your controller
+As the v0.4 framework accepts binary transactions itself, remove the endpoint that accepted
+transactions from the clients.
+
+### Use Java 11
+We have updated all project dependencies to the most recent versions supporting Java 11 and made
+our internal system tests run on 11. 8 is still fine too.
```
Comment (MakarovS) [inline doc/Migration_guide_0.4.md]: I believe users who run it on Java 8 (after we publish it built with Java 11) might encounter `java.lang.NoSuchMethodError: java.nio.ByteBuffer.rewind()Ljava/nio/ByteBuffer;` exception.

File: doc/Migration_guide_0.4.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+## Explore the new features
+
+### Use the Core Schema
+v0.4 provides access to some data stored by the framework in the database: transaction messages,
+execution results; blocks; node configuration, etc — take a look at `Blockchain` to see if will
```
Comment (MakarovS) [inline doc/Migration_guide_0.4.md]: `if it will simplify`?

File: doc/Migration_guide_0.4.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+## Questions?
+
+In case of any questions, feel free to [ask][gitter] a question in Gitter or [open][new-issue] an issue
+on Github.    
+
+[gitter]: https://gitter.im/exonum/exonum-java-binding
+[new-issue]: https://github.com/exonum/exonum-java-binding/issues/new
+
+
```
Comment (MakarovS) [inline doc/Migration_guide_0.4.md]: redundant line

File: doc/Migration_guide_0.4.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+## Update the client apps
+Update the clients to submit transactions in the new format, introduced in Exonum 0.10.
+The most notable change in the format is that the public key of the transaction author 
+is always included in the transaction message. This allows the framework to verify 
+the message signature with no conversion to an executable transaction.
+
+If the client application is in Java, use `com.exonum.binding.common.message.TransactionMessage`
+from "[com.exonum.binding:exonum-java-binding-common:0.4][common-0.4]";
+if it is in Java Script, use the latest version of the JS
```
Comment (vitvakatu) [inline doc/Migration_guide_0.4.md]:
```suggestion
if it is in Java Script, use the `0.16` version of the JS 
```

File: doc/Migration_guide_0.4.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
+# Migrating Java service to v0.4
+
+## Update the client apps
+Update the clients to submit transactions in the new format, introduced in Exonum 0.10.
+The most notable change in the format is that the public key of the transaction author 
+is always included in the transaction message. This allows the framework to verify 
+the message signature with no conversion to an executable transaction.
+
+If the client application is in Java, use `com.exonum.binding.common.message.TransactionMessage`
+from "[com.exonum.binding:exonum-java-binding-common:0.4][common-0.4]";
```
Comment (dmitry-timofeev) [inline doc/Migration_guide_0.4.md]: We might as well release `com.exonum.client:exonum-light-client:0.1` by that time :thinking:

File: exonum-java-binding/doc/Migration_guide_0.4.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+That leaves `Transaction` interface with two methods: `execute` and `info`.
+The argument of `execute` changed too — it now accepts a `TransactionContext` that allows
+to access a `Fork` to modify the database state, as previously; and some information about 
+the corresponding transaction message: its hash and the public key of the author.
+
+[blockchain-get-tx-messages-jd]: https://exonum.com/doc/api/java-binding-core/0.4/com/exonum/binding/blockchain/Blockchain.html#getTxMessages()
+[tx-context-hash-jd]: https://exonum.com/doc/api/java-binding-core/0.4/com/exonum/binding/transaction/TransactionContext.html#getTransactionMessageHash()
+
+### Update your controller
+As the v0.4 framework accepts binary transactions itself, remove the endpoint that accepted
```
Comment (bullet-tooth) [inline exonum-java-binding/doc/Migration_guide_0.4.md]:
```suggestion
As the v0.4 framework accepts binary transactions itself, remove the endpoints those accepted
```

File: exonum-java-binding/doc/Migration_guide_0.4.md [hunk 13]
```diff
@@ -0,0 +1,110 @@
+# Migrating Java service to v0.4
+
+## Update the client apps
+Update the clients to submit transactions in the new format, introduced in Exonum 0.10.
+The most notable change in the format is that the public key of the transaction author 
+is always included in the transaction message. This allows the framework to verify 
+the message signature with no conversion to an executable transaction.
+
+If the client application is in Java, 
+use [`com.exonum.binding.common.message.TransactionMessage`][tx-message-jd]
```
[the remark this answers is not in the document]
Comment (dmitry-timofeev) [reply exonum-java-binding/doc/Migration_guide_0.4.md]: Makes sense, thanks!

**Pull request merged**